toposortTyVarsOf doesn't handle foralls

When I originally wrote the code for `toposortTyVarsOf`, I left a bit of partiality in:

https://github.com/goldfirere/th-desugar/blob/1b76cf6da5a28ad133126aa77201080ff0d83f51/Language/Haskell/TH/Desugar/Core.hs#L1326-L1330

At the time, I thought that this was OK, since `toposortTyVarsOf` would only ever be called on type family patterns, which shouldn't ever contain `forall`s. But I later found _other_ uses for `toposortTyVarsOf`, such as in `singletons`, where I use it to properly sort the free variables in defunctionalization symbols' return kinds. And wouldn't you know it, those variables _can_ have polytypes appear in their kinds. As an example of this, the following piece of `singletons` code crashes:

```haskell
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE TemplateHaskell #-}
module Bug where

import Data.Singletons.TH

$(singletons [d|
  type Foo (x :: y) (f :: forall a. a) = f Bool
  |])
```
```
$ /opt/ghc/8.6.1/bin/ghc Bug.hs
[1 of 1] Compiling Bug              ( Bug.hs, Bug.o )

Bug.hs:1:1: error:
    Exception when trying to run compile-time code:
      `forall` type used in type family pattern
CallStack (from HasCallStack):
  error, called at ./Language/Haskell/TH/Desugar/Core.hs:1330:30 in th-desugar-1.9-IWJeiZlphUgItyl5W2YZ4A:Language.Haskell.TH.Desugar.Core
    Code: singletons
            [d| type Foo (x :: y) (f :: forall a. a) = f Bool |]
  |
1 | {-# LANGUAGE PolyKinds #-}
  | ^
```

(Not that I expect that this code _should_ typecheck necessarily, since trying to defunctionalize a type synonym with a higher-ranked argument like this has other problems. But at the very least, it shouldn't crash.)

In light of this, it seems that we should just bite the bullet and implement the `DForallT` case of `toposortTyVarsOf` properly. Here is a minimal test case which triggers the `error`:

```haskell
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE TemplateHaskell #-}
module Bug where

import Language.Haskell.TH
import Language.Haskell.TH.Desugar

$(do let a = mkName "a"
     t  <- [t| forall f. f $(varT a) |]
     t' <- desugar t
     runIO $ print $ toposortTyVarsOf [t']
     pure [])
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,6 +1,11 @@
 `th-desugar` release notes
 ==========================
 
+Version next
+------------
+* Fix a bug in which `toposortTyVarsOf` would error at runtime if given types
+  containing `forall`s as arguments.
+
 Version 1.9
 -----------
 * Suppose GHC 8.6.

--- a/Language/Haskell/TH/Desugar/Core.hs
+++ b/Language/Haskell/TH/Desugar/Core.hs
@@ -1324,21 +1324,38 @@ toposortTyVarsOf tys =
       fvs = Set.toList $ foldMap fvDType tys
 
       varKindSigs :: Map Name DKind
-      varKindSigs = foldMap go tys
+      varKindSigs = foldMap go_ty tys
         where
-          go :: DType -> Map Name DKind
-          go (DForallT {}) = error "`forall` type used in type family pattern"
-          go (DAppT t1 t2) = go t1 `mappend` go t2
-          go (DSigT t k) =
-            let kSigs = go k
+          go_ty :: DType -> Map Name DKind
+          go_ty (DForallT tvbs ctxt t) =
+            go_tvbs tvbs (foldMap go_pred ctxt `mappend` go_ty t)
+          go_ty (DAppT t1 t2) = go_ty t1 `mappend` go_ty t2
+          go_ty (DSigT t k) =
+            let kSigs = go_ty k
             in case t of
                  DVarT n -> Map.insert n k kSigs
-                 _       -> go t `mappend` kSigs
-          go (DVarT {}) = mempty
-          go (DConT {}) = mempty
-          go DArrowT    = mempty
-          go (DLitT {}) = mempty
-          go DWildCardT = mempty
+                 _       -> go_ty t `mappend` kSigs
+          go_ty (DVarT {}) = mempty
+          go_ty (DConT {}) = mempty
+          go_ty DArrowT    = mempty
+          go_ty (DLitT {}) = mempty
+          go_ty DWildCardT = mempty
+
+          go_pred :: DPred -> Map Name DKind
+          go_pred (DForallPr tvbs ctxt pr) =
+            go_tvbs tvbs (foldMap go_pred ctxt `mappend` go_pred pr)
+          go_pred (DAppPr pr ty) = go_pred pr `mappend` go_ty ty
+          go_pred (DSigPr pr ki) = go_pred pr `mappend` go_ty ki
+          go_pred (DVarPr{})     = mempty
+          go_pred (DConPr{})     = mempty
+          go_pred DWildCardPr    = mempty
+
+          go_tvbs :: [DTyVarBndr] -> Map Name DKind -> Map Name DKind
+          go_tvbs tvbs m = foldr go_tvb m tvbs
+
+          go_tvb :: DTyVarBndr -> Map Name DKind -> Map Name DKind
+          go_tvb (DPlainTV n)    m = Map.delete n m
+          go_tvb (DKindedTV n k) m = Map.delete n m `mappend` go_ty k
 
       (g, gLookup, _)
         = graphFromEdges [ (fv, fv, kindVars)

--- a/Test/Run.hs
+++ b/Test/Run.hs
@@ -289,6 +289,13 @@ test_t85 =
        expanded_ty <- expandType orig_ty
        expected_ty `eqTHSplice` expanded_ty)
 
+test_t92 :: Bool
+test_t92 =
+  $(do a <- newName "a"
+       f <- newName "f"
+       let t = DForallT [DPlainTV f] [] (DVarT f `DAppT` DVarT a)
+       toposortTyVarsOf [t] `eqTHSplice` [DPlainTV a])
+
 test_getDataD_kind_sig :: Bool
 test_getDataD_kind_sig =
 #if __GLASGOW_HASKELL__ >= 800
@@ -465,6 +472,8 @@ main = hspec $ do
     it "doesn't crash on a stuck type family application" $ test_stuck_tyfam_expansion
 
     it "expands type synonyms in kinds" $ test_t85
+
+    it "toposorts free variables in polytypes" $ test_t92
 
     it "reifies data type return kinds accurately" $ test_getDataD_kind_sig
 

